File: leetcode-cn/20.有效的括号.py
#
# @lc app=leetcode.cn id=20 lang=python3
#
# [20] 有效的括号
#


# @lc code=start
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []

        for c in s:
            if c == ')':
                if stack and stack[-1] == '(':
                    stack.pop(-1)
                else:
                    return False
            elif c == ']':
                if stack and stack[-1] == '[':
                    stack.pop(-1)
                else:
                    return False
            elif c == '}':
                if stack and stack[-1] == '{':
                    stack.pop(-1)
                else:
                    return False
            else:
                stack.append(c)

        if not stack:
            return True
        else:
            return False


# @lc code=end

File: leetcode-cn/C++/20.有效的括号.cpp
/*
 * @lc app=leetcode.cn id=20 lang=cpp
 *
 * [20] 有效的括号
 */

// @lc code=start
#include <stack>
#include <string>
#include <unordered_map>

using namespace std;

class Solution
{
public:
    bool isValid(string s)
    {
        unordered_map<char, char> map = {
            {')', '('},
            {'}', '{'},
            {']', '['},
        };
        stack<char> st;

        for (char c : s)
        {
            if (c == '(' || c == '{' || c == '[')
            {
                st.push(c);
            }
            else
            {
                if (st.empty() || map[c] != st.top())
                {
                    return false;
                }
                st.pop();
            }
        }

        if (st.empty())
        {
            return true;
        }

        return false;
    }
};
// @lc code=end
